crates/env/src/call/create_builder.rs

```diff
@@ -339,54 +395,140 @@ where
         GasLimit,
         Set<E::Balance>,
         Set<ExecutionInput<Args>>,
         Set<Salt>,
-        R,
+        Set<ReturnType<RetType>>,
     >
 where
     E: Environment,
     GasLimit: Unwrap<Output = u64>,
 {
-    /// Sets the value transferred upon the execution of the call.
+    /// Finalizes the create builder, allowing it to instantiate a contract.
     #[inline]
-    pub fn params(self) -> CreateParams<E, Args, Salt, R> {
+    pub fn params(self) -> CreateParams<E, Args, Salt, RetType> {
         CreateParams {
             code_hash: self.code_hash.value(),
             gas_limit: self.gas_limit.unwrap_or_else(|| 0),
             endowment: self.endowment.value(),
             exec_input: self.exec_input.value(),
             salt_bytes: self.salt.value(),
-            _return_type: self.return_type,
+            _return_type: Default::default(),
         }
     }
 }
 
-impl<E, GasLimit, Args, Salt, R>
+impl<E, GasLimit, Args, Salt, RetType>
     CreateBuilder<
         E,
         Set<E::Hash>,
         GasLimit,
         Set<E::Balance>,
         Set<ExecutionInput<Args>>,
         Set<Salt>,
-        R,
+        Set<ReturnType<RetType>>,
     >
 where
     E: Environment,
     GasLimit: Unwrap<Output = u64>,
     Args: scale::Encode,
     Salt: AsRef<[u8]>,
-    R: FromAccountId<E>,
+    RetType: FromAccountId<E>,
 {
     /// Instantiates the contract using the given instantiation parameters.
+    ///
+    /// # Panics
+    ///
+    /// This method panics if it encounters an [`ink_primitives::LangError`]. If you want to handle
+    /// those use the [`try_instantiate`][`CreateBuilder::try_instantiate`] method instead.
     #[inline]
-    pub fn instantiate(self) -> Result<R, Error> {
+    pub fn instantiate(self) -> Result<RetType, Error> {
         self.params().instantiate()
     }
+
+    /// Instantiates the contract using the given instantiation parameters.
+    ///
+    /// # Note
+    ///
+    /// On failure this returns an [`ink_primitives::LangError`] which can be handled by the caller.
+    #[inline]
+    pub fn try_instantiate(
+        self,
+    ) -> Result<::ink_primitives::ConstructorResult<RetType>, Error> {
+        self.params().try_instantiate()
+    }
+}
+
+impl<E, GasLimit, Args, Salt, RetType, ContractError>
+    CreateBuilder<
+        E,
+        Set<E::Hash>,
+        GasLimit,
+        Set<E::Balance>,
+        Set<ExecutionInput<Args>>,
+        Set<Salt>,
+        Set<ReturnType<::core::result::Result<RetType, ContractError>>>,
+    >
+where
+    E: Environment,
+    GasLimit: Unwrap<Output = u64>,
+    Args: scale::Encode,
+    Salt: AsRef<[u8]>,
+    RetType: FromAccountId<E>,
+    ContractError: scale::Decode,
+{
+    /// Instantiates the contract using the given instantiation parameters.
+    ///
+    /// # Note
+    ///
+    /// On failure this returns an [`ink_primitives::LangError`] which can be handled by the caller.
+    #[inline]
+    pub fn try_instantiate_fallible(
```

Comment:
Ideally we don't have this as part of the `CreateBuilder`, and instead we allow the implementation to dispatch the correct method based off the concrete types used (so whatever is set with `returns::<T>()`. 

I haven't been able to figure out a way to do this yet, @ascjones you got any ideas?

Reply:
I might have an idea for how to handle this, currently verifying whether it will work or not

Reply:
*WIP* here: https://github.com/paritytech/ink/pull/1591

Reply:
Ok the PR for my idea basically works now: https://github.com/paritytech/ink/pull/1591. So please have a look and let me know what you think.